github-api.php

```diff
@@ -3201,71 +3201,388 @@ function vipgoci_github_prs_commits_list(
-			'from GitHub' .
+			'from git repository' .
 			vipgoci_cached_indication_str( $cached_data ),
 
 		array(
-			'repo_owner' => $repo_owner,
-			'repo_name' => $repo_name,
-			'commit_id_a' => $commit_id_a,
-			'commit_id_b' => $commit_id_b,
+			'local_git_repo'	=> $local_git_repo,
+			'commit_id_a'		=> $commit_id_a,
+			'commit_id_b'		=> $commit_id_b,
 		)
 	);
 
 	if ( false === $cached_data ) {
 		/*
-		 * Nothing cached; ask GitHub.
+		 * Nothing cached; look in git repository.
 		 */
 
-		// FIXME: Use local git-repo for this, if possible.
-
-		$github_url =
-			VIPGOCI_GITHUB_BASE_URL . '/' .
-			'repos/' .
-			rawurlencode( $repo_owner ) . '/' .
-			rawurlencode( $repo_name ) . '/' .
-			'compare/' .
-			rawurlencode( $commit_id_a ) .
-				'...' .
-				rawurlencode( $commit_id_b );
+		$git_diff_cmd = sprintf(
+			'%s -C %s diff %s %s 2>&1',
+			escapeshellcmd( 'git' ),
+			escapeshellarg( $local_git_repo ),
+			escapeshellarg( $commit_id_a ),
+			escapeshellarg( $commit_id_b )
+		);
 
-		// FIXME: Error-handling
-		$resp_raw = json_decode(
-			vipgoci_github_fetch_url(
-				$github_url,
-				$github_token
+		vipgoci_log(
+			'Running git...',
+			array(
+				'cmd'	=> $git_diff_cmd
 			)
 		);
 
+		/* Actually execute */
+		vipgoci_runtime_measure( VIPGOCI_RUNTIME_START, 'git_cli' );
+
+		$git_diff_results = shell_exec( $git_diff_cmd );
+
+		vipgoci_runtime_measure( VIPGOCI_RUNTIME_STOP, 'git_cli' );
+
+		/*
+		 * Prepare intermediate results array.
+		 */
+		$diff_results = array(
+			'files' => array()
+		);
+
+		/*
+		 * Split results into array
+		 */
+
+		$git_diff_results = explode(
+			PHP_EOL,
+			$git_diff_results
+		);
+
+		/*
+		 * Initialize stateful variables.
+		 */
+
+		$cur_file = null;
+		$cur_mode = 'info'; // Other mode is 'patch'
+		$cur_file_first_patch_line = true;
+		$cur_file_status = null;
+		$cur_file_minus = null;
+		$cur_file_plus = null;
+		$cur_file_paths_cleaned = false;
+		$cur_file_plus_plus_plus_seen = false;
+		$cur_file_patch_buffer = '';
+
+		/*
+		 * Process the output of git diff, one line
+		 * at a time. Update state-variables as we go
+		 * so we can keep track of what is happening.
+		 *
+		 * The output looks like this:
+		 * 
+		 *  diff --git a/renamed-file2.txt b/renamed-file2.txt
+		 *  deleted file mode 100644
+		 *  index 55ab87b..0000000
+		 *  --- a/renamed-file2.txt
+		 *  +++ /dev/null
+		 *  @@ -1,2 +0,0 @@
+		 *  -# vip-go-ci-testing
+		 *  -Pull-Requests, commits a
+		 *  diff --git ...
+		 *  ...
+		 */
+
+		foreach( $git_diff_results as $git_result_item ) {
+			/*
+			 * Split each line into array at spaces,
+			 * making it easy to process each line of results.
+			 */
+			$git_result_item_arr = explode(
+				' ',
+				$git_result_item
+			);
+
+			/*
+			 * Check if we are seeing 'diff --git ... 
+			 * and if so switch to info mode, get file names
+			 * and set other state variables.
+			 */
+			if (
+				( ! empty( $git_result_item_arr ) ) &&
```

Comment:
I think I would keep this as is to avoid another indent. I sometimes find it more difficult to read the code when there the number of indents have become many. We already have four levels in on line 3330 for example.